### ISSUE TITLE
Status, timestamp field (sTs), what should this value reflect.

### ISSUE THREAD

**Issue by sveitech** [rsmp-nordic/rsmp_core#122]
From the documents: https://rsmp-nordic.github.io/rsmp_specifications/core/3.2/applicability/basic_structure.html#status-messages

section 4.4.4.1.2:

> Timestamp for the status. The timestamp uses the W3C XML dateTime definition with a 3 decimal places. All timestamps are set at the site (and not in the supervision system) when the status is fetched (and not when the message is sent) All timestamps uses UTC.

What does this mean? What does "when the status is fetched" mean? fetched from where?

Should the timestamp reflect when the status field changed its value? Our should it reflect when it was queried by the server?
How should this timestamp be treated for subscriptions (update-rate > 0)? Should it still only reflect when the value actually changed?

In our current implementation (8 years old now) the timestamps are set to the time at which we send the package to the server, not when the field changed (for update-rate subscriptions). This seems wrong now, given this description in the documents.

**Comment by otterdahl** [comment]
> What does this mean? What does "when the status is fetched" mean? fetched from where?
> 
> Should the timestamp reflect when the status field changed its value?

Not exactly. Typically, you don't need to keep track exactly when each value changes unless there is an active subscription with update-on-change.

> Our should it reflect when it was queried by the server?

Yes. The timestamp should reflect when the current value was queried - which is typically occurs when it was requested by the server or by any active subscription.

But it should not reflect when the message was sent, because then the time stamp could be affected by network delay or connection outage. If this was the case than it could be interpreted that the timestamp for the buffered messages are updated when sent during connection handshake. But if that was true, then the buffered data wouldn't be of much use for statistics. We like to know for how long each status has been active, regardless if there has been any connectivity delays or outages issues.

> How should this timestamp be treated for subscriptions (update-rate > 0)? Should it still only reflect when the value actually changed?

Only when the update-rate is "update on change". Otherwise it is reflects the interval of the subscription.

We understand that values may sometimes update very frequently (e.g. signal group status). But since the intention is that the time stamp should reflect when the value was queried, the idea is that time stamps shouldn't include the network delay and it should be possible to e.g. log the actual duration of each status for use in statistics.

**Comment by emiltin** [comment]
> The timestamp should reflect when the current value was queried - which is typically occurs when it was requested by the server or by any active subscription.

The way I understand this is that it's the time the value is determined at the site, not when it changes. So in practical terms:

1. when you receive a status request
2. when you receive a status subscription
3. when a subscription updateRate triggers
4. when a subscription sendOnChange triggers

> Yes. The timestamp should reflect when the current value was queried - which is typically occurs when it was requested by the server or by any active subscription.

If network is OK, then the site determines the value and sends it immediately. But if connection is down, it determines the value, then have to buffer it and send it later. In this case the timestamp should reflect when the value is determined, not send.

The terminology is tricky. Unfortunately both 'read', 'fetch' and 'query' could be understood as the server requesting the data.

**Comment by emiltin** [comment]
Here's how I see a site handles each type of event. In all cases, it essentially read that value and uses the current time as the timestamp. You never need to keep track of when a value changed:

**StatusRequest received**
read the value and store the current time as timestamp
send a status response using that timestamp

**StatusSubscription receive**
update subscriptions
if the not already subscribed:
read the value and store the current time as timestamp
send a status update using that timestamp

**Subscription updateRate triggers**
read the value and store the current time as timestamp
send a status update using that timestamp

**Subscription sendOnChange triggers**
read the value and store the current time as timestamp
send a status update using that timestamp
